app/main/api/handlers/getState.ts

```diff
@@ -0,0 +1,11 @@
+import { GetStateParams, GetStateResponse } from "app/common/runtimeTypes/ipc/state"
+import { SubSystems } from "app/main/system"
+
+/**
+ * AppInfo handler
+ */
+async function getState(system: SubSystems, params: GetStateParams): Promise<GetStateResponse> {
```

Comment:
I don't get how could `GetStateParams` work here since params is always an `JsonSerializable` (this actually must be fixed because I got it wrong -I'll open an issue-). There's no way handlers can get their params deserialized because only the handler knows to which business type the params must be deserialized to. So each handler will always receive a `params: JsonSerializable` and will have to deserialize that into the appropriate business type.

This is how each handler should look like:
```
              +---------------------------------------+
    request   |  +-------------------+                |
  ------------+-->  asRuntimeType    |                |
              |  +-------|-----------+                |
              |  +-------v-----------+                |
              |  | logic that works  |                |
              |  | with complex      |                |
              |  | business type     |                |
              |  +-------|-----------+                |
    response  |  +-------v-----------+                |
  <-----------+--+  asObject         |                |
              |  +-------------------+                |
              |                                       |
              |                          a handler    |
              |                                       |
              +---------------------------------------+
```

There's actually an automatic way of doing this by using decorators and reflect-metadata but that would force us to write the handlers as methods in a class because decorators won't work with plain functions.

Reply:
I also thought of decorators when I came with this, but I found the same roadblock as you mentioned: they can only be used on class methods because of weird JS ins and outs.